Fix conversion error shallowed and normalize conversion status

ivanortegaalba opened the pull request:
**Summary**

- Fix conversion error handling to properly report metrics and logs by standardizing all conversion functions to return errors
- Add data loss detection for all version permutations (v0alpha1, v1beta1, v2alpha1, v2beta1)
- Add comprehensive regression prevention tests
- Fix type mismatch bug in query counting for V2 to V0/V1 dashboard conversions
- Remove unused `withConversionDataLossDetection` function (dead code)

**Problem**

Two different error-handling patterns existed in conversion functions:

**Pattern A (Correct)**: Returns error for metrics/logging
```go
if err != nil {
    out.Status.Conversion.Failed = true
    return err  // Metrics wrapper sees the error
}
```

**Pattern B (Incorrect)**: Swallows error
```go
if err != nil {
    out.Status.Conversion.Failed = true
    return nil  // Metrics wrapper thinks conversion succeeded
}
```

When Pattern B swallowed errors:
1. `withConversionMetrics` received `nil` and assumed success
2. It ran `checkConversionDataLoss()` on a partially converted dashboard
3. The target dashboard had empty/nil spec (because conversion failed internally)
4. Data loss was detected (comparing source panels vs target's zero panels)
5. Result: False `conversion_data_loss_error` metrics instead of actual conversion errors

***Additional issue - Type mismatch in query counting:***

The conversion function in `v2alpha1_to_v1beta1.go` created targets as:
```go
targets := make([]map[string]interface{}, 0, len(spec.Data.Spec.Queries))
```

But the counting function expected `[]interface{}` (from JSON unmarshaling). The type assertion failed silently, returning 0 queries and triggering false data loss errors.

**Production symptoms:**
- Success metrics stopped reporting around December 3rd for certain conversion paths
- `conversion_data_loss_error` metrics increased unexpectedly
- Real conversion errors were hidden
- No error logs appeared for conversion failures

**Solution**

1. **Standardize on Pattern A**: All conversion functions now return errors properly
2. **Data loss check timing**: Only runs when conversion succeeds (`err == nil`)
3. **Metrics wrapper unchanged**: Still returns `nil` to API server to prevent 500 errors
4. **Add success status**: All conversions now set `Status.Conversion.Failed=false` on success
5. **Add consistency fixes**: ObjectMeta/APIVersion set on error, default layout for v2 targets on error
6. **Fix type mismatch**: Changed `v2alpha1_to_v1beta1.go` to create targets as `[]interface{}` for consistency

**How to test**

Run conversion tests:
```
go test ./apps/dashboard/pkg/migration/conversion/...
```

Run the new integration tests:
```
go test -v -run "TestDataLossDetection_IntegrationWithRealConversions" ./apps/dashboard/pkg/migration/conversion/
```

Verify metrics after deployment:
```
sum by (error_type) (rate(grafana_dashboard_conversion_failure_total[5m]))
```
Should see increase in conversion_error, decrease in conversion_data_loss_error

Verify logs:
```
{app="grafana"} |= "Dashboard conversion failed" | json | errorType="conversion_error"
```
```

ivanortegaalba committed:
Fix conversion error shallowed and normalize conversion status

ivanortegaalba committed:
Add unit to ensure all permutations dataloss detection

ivanortegaalba committed:
Update apps/dashboard/pkg/migration/conversion/metrics.go

ivanortegaalba committed:
Remove rebundant test

ivanortegaalba committed:
refactor assertion

ivanortegaalba committed:
avoid ifs statements

dprokop commented on apps/dashboard/pkg/migration/conversion/v2.go:
If we return error here, would the response with `status` set be returned to the client?

dprokop replied:
nvm - figured this out!

dprokop reviewed (approved):
Looks good, just a tiny thing that you can cleanup.

ivanortegaalba committed:
Merge remote-tracking branch 'origin/main' into iortega/fix-metrics-and-status-inconsistencies

ivanortegaalba committed:
Fix counting issue

ivanortegaalba commented:
I also fixed a type mismatch bug in query counting for V2 to V0/V1 dashboard conversions.

**Problem**
The conversion function created targets as []map[string]interface{} but the counting function expected []interface{} (which is what JSON unmarshaling produces). This caused the type assertion to fail silently, reporting 0 queries and triggering false positive data loss errors.

**Changes**
- Changed v2alpha1_to_v1beta1.go to create targets as []interface{} 
- Added countTargetsFromPanel helper to centralize target counting
- Added integration tests with real scheme.Convert calls covering all layout types

ivanortegaalba committed:
Merge remote-tracking branch 'origin/main' into iortega/fix-metrics-and-status-inconsistencies

ivanortegaalba committed:
Adapt to main

ivanortegaalba committed:
Fix linter

ivanortegaalba committed:
Merge remote-tracking branch 'origin/main' into iortega/fix-metrics-and-status-inconsistencies

# Conflicts:
#	apps/dashboard/pkg/migration/conversion/conversion_data_loss_detection.go
#	apps/dashboard/pkg/migration/conversion/testdata/output/migrated_dev_dashboards/panel-gauge/v0alpha1.gauge_tests_new.v42.v2alpha1.json
#	apps/dashboard/pkg/migration/conversion/testdata/output/migrated_dev_dashboards/panel-gauge/v0alpha1.gauge_tests_new.v42.v2beta1.json
#	apps/dashboard/pkg/migration/conversion/testdata/output/migrated_dev_dashboards/panel-gauge/v0alpha1.gauge_tests_old_to_new.v42.v2alpha1.json
#	apps/dashboard/pkg/migration/conversion/testdata/output/migrated_dev_dashboards/panel-gauge/v0alpha1.gauge_tests_old_to_new.v42.v2beta1.json
